lib/utils/fanoutbuffer/buffer.go

```diff
@@ -0,0 +1,426 @@
+	overflowStart time.Time
+
+	// overflow is a variable length buffer of items that still need to be observed.
+	overflow []entry[T]
+}
+
+// entry represents an item in a buffer
+type entry[T any] struct {
+	item T
+	// wait is the number of cursors that still need to observe this item. each cursor
+	// decrements this value when it observes the item. items with wait==0 are cleaned
+	// up. Note that this atomic is only safe to decrement while read lock is held. Entries
+	// may be moved while write lock is held.
+	wait atomic.Uint64
+}
+
+// NewBuffer creates a new fanout buffer instance.
+func NewBuffer[T any](cfg Config) *Buffer[T] {
+	cfg.SetDefaults()
+	return &Buffer[T]{
+		cfg:    cfg,
+		notify: make(chan struct{}),
+		ring:   make([]entry[T], cfg.Capacity),
+	}
+}
+
+// Close permanently closes the fanout buffer. Note that all cursors are terminated immediately
+// and may therefore miss items that had not been read at the time Close is called.
+func (b *Buffer[T]) Close() {
+	b.write(func() {
+		b.closed = true
+		close(b.notify)
+	})
+}
+
+// NewCursor gets a new cursor into the buffer. Cursors *must* be closed as soon as they are
+// no longer being read from. Stale cursors may cause performance degredation.
+func (b *Buffer[T]) NewCursor() *Cursor[T] {
+	var c uint64
+	b.write(func() {
+		if b.closed {
+			return
+		}
+		c = b.pos + b.len
+		b.cursors++
+	})
+
+	cursor := &Cursor[T]{
+		buf:    b,
+		cursor: c,
+	}
+	runtime.SetFinalizer(cursor, finalizeCursor[T])
+	return cursor
+}
+
+// Append appends to the buffer and wakes all dormant cursors.
+func (b *Buffer[T]) Append(items ...T) {
+	var notify chan struct{}
+
+	b.write(func() {
+		if b.closed {
+			panic("Append called on closed fanout buffer")
+		}
+
+		// free up slots that are no longer needed
+		b.cleanupSlots()
+
+		for _, item := range items {
+			// ensure that there is at least one free slot in ring
+			b.ensureSlot()
+
+			// insert item into next free slot in ring
+			idx := int((b.pos + b.len) % b.cfg.Capacity)
+			b.ring[idx].item = item
+			b.ring[idx].wait.Store(b.cursors)
+			b.len++
+		}
+
+		notify = b.notify
+		b.notify = make(chan struct{})
+	})
+
+	// notify all waiting cursors. we do this outside of the
+	// lock to (theoretically) reduce potential contention.
+	close(notify)
+}
+
+// cleanupSlots frees up enries that have been seen by all cursors and clears the backlog if
+// it has exceeded its grace period.
+func (b *Buffer[T]) cleanupSlots() {
+	// reset overflow if it is past its grace period
+	if !b.overflowStart.IsZero() && b.cfg.Clock.Now().After(b.overflowStart.Add(b.cfg.GracePeriod)) {
+		b.overflow = nil
+	}
+
+	// trim items from overflow that have been seen by all cursors
+	var clearOverflowTo int
+	for i := 0; i < len(b.overflow); i++ {
+		clearOverflowTo = i
+		if b.overflow[i].wait.Load() > 0 {
+			break
+		}
+	}
+	b.overflow = append(b.overflow[:0], b.overflow[clearOverflowTo:]...)
```

Comment:
What's the reason for using append instead of just:
```suggestion
	b.overflow = b.overflow[clearOverflowTo:]
```

Coming back to this comment later in review... is this meant to prevent the underlying array from growing indefinitely?

Reply:
The underlying array wouldn't grow indefinitely because if you append to it and there's no more capacity, only the "live" items will get copied over a new array anyway; in hindsight doing that is probably lighter than moving all the slice back one by one.